[team-07] [Hwi & 땃쥐] issuetracker 4차 PR

ttasjwi opened the pull request:
안녕하세요 디온, 땃쥐 & 휘입니다!!
주문하신 스파게티가 나왔습니닷...!!!
주문이 늦어져서 죄송합니다ㅠㅠ 대신 미트볼을 추가해드렸습니다...:sob::sob::sob:

---

## 💻 주요 작업 내역

+ 이슈 생성 리팩토링
+ 이슈 수정
+ 이슈 다중상태 변경 최적화 : 쿼리 한번으로 하도록 함
+ 라벨 등록
+ 등록 시간과 수정 시간을 자동 갱신
+ 상세 조회 최적화

---

### 생성, 수정시각 공통화
![image](https://user-images.githubusercontent.com/88282356/175479489-b12d5631-15da-405b-b18c-a27f2cd0b30f.png)
- JPA Auditing을 활용하여 생성시각, 날짜를 공통화 처리했습니다.

---

### 상세조회 최적화
```java=
    @Override
    public Optional<IssueDetailResponse> findIssueDetail(Long issueId) {

        IssueDetailResponse issueDetailResponse = getIssueDetailResponse(issueId);

        if (issueDetailResponse == null) {
            return Optional.empty();
        }

        List<IssueDetailResponse.IssueAssigneeDto> assigneeDtos = findIssueAssigneeDtos(issueId);
        List<LabelDto> labelDtos = findLabelDtos(issueId);
        List<IssueDetailResponse.CommentListElement> commentDtos = findCommentDtos(issueId);

        issueDetailResponse.getAssignees().addAll(assigneeDtos);
        issueDetailResponse.getLabels().addAll(labelDtos);
        issueDetailResponse.getComments().addAll(commentDtos);

        return Optional.of(issueDetailResponse);
    }
```
- `IssueRepositoryImpl`에서 이슈 상세 DTO를 바로 조회하도록 함


---

## 지난번 PR
- 시간 관계상 모든걸 반영하지는 못 했습니다. ㅜㅜ
- 당장 디온이 말씀 주셨던 상세조회 부분을 다시 보니 매우 심각한 느낌이 들었습니다. 이 부분을 중점으로 했습니다.

---

## 다음에 할 것
- 마저 못 한 기능 구현
- OAuth/JWT
    - 우선순위 매우 높음. 이번에 학습하고 싶었는데 자꾸 미뤄서 아쉬웠어요. 좀 힘들고 시행착오가 있을 수 있겠지만 이번 김에 제대로 배워보고 싶습니다..)
- 예외처리
    - ExceptionHandler, Exception 메시지 공통화 처리
- 인터셉터 처리
  - 해당 요청에 대해 사용자가 접근 권한이 있는지 컨트롤러 접근 전에 판단하는 로직 (JWT, OAuth 선행 필요)
- 테스트 : 그동안 시간을 핑계로 테스트를 너무 미뤘습니다... 다음 주 수료식 일정 등등으로 인해 이 부분까지 할 수 있으리란 보장이 없긴 합니다. ㅜㅜ

ttasjwi committed:
[wnsxor1993#71] refactor: issue 생성 리팩토링

- 기존에는 issue의 create 메서드에서 생성, 할당자 부착, 라벨 부착을 모두 책임졌다.
- create와 changeAssignee, changeLabel을 분리하기로 함.
- 할당자 변경 기능 구현

ttasjwi committed:
[wnsxor1993#72] feat: 스프링 데이터 Auditing를 적용하여 등록 시간과 수정 시간을 자동 갱신

ttasjwi committed:
[wnsxor1993#83] refactor: Issue 클래스의 1:多 관계 제어자 final로 변경

ttasjwi committed:
[wnsxor1993#73] style: 클래스파일 리포맷 작업

- 들여쓰기 등등 기본적인 ide의 정렬 기능을 이용해 코드 라인들을 정렬함.

hwicode committed:
[wnsxor1993#71] refactor: 이슈에 라벨을 붙히는 기능 리팩토링

ttasjwi committed:
[wnsxor1993#75] feat: issue 클래스 변경 메서드 추가

- changeTitle : 제목 변경
- changeContent : 본문 변경
- changeMilestone : 마일스톤 변경

ttasjwi committed:
[wnsxor1993#75] feat: 이슈 수정기능 추가

ttasjwi committed:
[wnsxor1993#76] move: DTO 패키지 내부 분류 작업

- DTO 패키지 하위에 모든 DTO가 모여있어서 이를 하위에 내부 분류함

ttasjwi committed:
[wnsxor1993#77] fix: IssueRepository에 있던 사용자 정의 메서드를 사용자 정의 레포지토리로 이동

- IssueRepository에 사용자정의 인터페이스를 구현해뒀는데 다시 보니 IssueCustomRepository에 정의해두지 않았음.
- 이동함

ttasjwi committed:
[wnsxor1993#63] refactor: IssueCommandService의 changeStates 최적화

- 단일 엔티티가 아닌 다중 엔티티들의 상태를 일괄적으로 변경해야한다.
- 근데 엔티티 하나하나 일일이 수정하는 것은 엔티티의 갯수만큼 쿼리가 나간다.
- 쿼리 한 방으로 In절에 속한 엔티티들의 상태를 변경하는게 더 DB 통신 비용이 저렴하다면 그것을 쓰는게 낫지 않을까?
- 한방 쿼리로 이슈들의 isOpened를 변경하기

ttasjwi committed:
[wnsxor1993#78] feat: 라벨 서비스 계층 추가

- 라벨 서비스 계층 추가 : QueryService, CommandService

ttasjwi committed:
[wnsxor1993#78] feat: 라벨 생성 요청, 응답 매핑 객체 생성

- 요청 매핑 : LabelCreateRequest
- 응답 매핑 : LabelCreateResponse

ttasjwi committed:
[wnsxor1993#78] feat: LabelCommandService - 라벨 등록 메서드 구현

ttasjwi committed:
[wnsxor1993#78] feat: LabelController 정의 및 라벨 추가 메서드 작성

- createLabel : 라벨 생성 요청을 받고 라벨 생성, 등록

ttasjwi committed:
[wnsxor1993#79] refactor: 이슈 단건 상세 조회 최적화

- ``@QueryProjection` : Querydsl에서 Dto 바인딩 시 사용할 수 있는 수단. 생성자에 달아주면 Dto의 Q타입을 만들어줌.
- 컴파일 단계에서 오류를 잡을 수 있고, IDE의 도움을 받아 쿼리를 작성하기 편하다.
- 단점 : DTO가 Querydsl를 알아야하는 상황이 발생하고, 별도로 QType을 만들어야함.

ttasjwi committed:
[wnsxor1993#81] refactor: IssueStateChange에서 상태 변경 후 issue를 다시 재질의 하지 않도록 함

- IssueStateChange에서 상태 변경 후 issue를 다시 재 질의 하지 않도록 함.
- 요청에 대해 응답 시, 그대로 issueId, isOpened를 그대로 반환
- IssueQueryService에서는 findIssue 메서드가 더 이상 public으로 공개될 필요도 없고 현재로서는 필요 없다. 제거.

ttasjwi committed:
[wnsxor1993#82] refactor: 컨트롤러 일부 메서드에 응답 Http 상태코드 지정

- `@ResponseStatus` 활용

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/domain/Comment.java:
```diff
@@ -34,10 +28,8 @@ public class Comment {
-    public Comment(String content, LocalDateTime createdAt, LocalDateTime lastModifiedAt, Member author, Issue issue) {
+    public Comment(String content, Member author, Issue issue) {
         this.content = content;
-        this.createdAt = createdAt;
-        this.lastModifiedAt = lastModifiedAt;
         this.author = author;
         this.issue = issue;
     }
```
점점 생성자가 관리하는 필드가 적어지는군요. 의도한건가요?

ttasjwi replied:
auditing으로 생성일자 수정일자를 자동 갱신해주기에 엔티티 코드에서 직접 관리하지 않도록 했습니다.

ksundong replied:
너무 JPA에 의존적인 코드가 되어가고 있진 않나요?

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/domain/Issue.java:
```diff
@@ -3,89 +3,167 @@
+    public void changeIssueAssignees(List<Member> changeMembers) {
+        removeAssigneesNotIn(changeMembers);
+        removeDuplicateMemberFrom(changeMembers);
 
-    private static void initIssueLabels(List<Label> labels, Issue issue) {
-        labels.forEach(label -> putLabelToIssue(label, issue));
+        List<IssueAssignee> newAssignees = createNewAssignees(changeMembers);
```
이런식으로 private 메서드로 쪼개는건 좋은 방식이 아닙니다.
왜 안좋을까요?

ttasjwi replied:
저희는 외부에서 changeIssueAssignees 메서드만 알게 하면 된다고 생각했고, 그 내부 로직은 외부에서 굳이 알게 할 필요가 없다고 생각해서 저 메서드들을 private로 만들었는데요.

디온의 의견을 듣고 이 구조의 문제가 무엇일까 논의해봤을 때, 저희는 이렇게 했을 경우 "테스트의 단위가 너무 커지겠다"라는 결론을 내게 됐습니다. 이를 반영하게 된다면 private 메서드들을 public으로 공개하면

- 각각의 작은 메서드 단위로 테스트할 수 있다. 각각의 메서드별로 테스트가 더 쉬워지고, 기능의 신뢰성이 증가된다.
- 다만, 외부에서는 대신 기존에 changeAssignees를 호출하던 것을 잘게 자른 메서드들을 순서대로 호출하는 식으로 처리해야한다.

이런 결론을 낼 수 있었는데요. 디온이 생각하신 의도와 저희 생각이 비슷한 것인지 궁금합니다. 혹시 다른 의도셨다면 그 의도를 확인할 수 있을까요?

ksundong replied:
네 의도와 전혀 다릅니다.

추출한 의도가 무엇인가요? 이건 외부에서 알지 못하게 하기 위한 추출과는 다르다고 보입니다.
오히려 로직을 블럭으로 나눠놓은 형태가 아닐까요?
그리고, 메서드는 기본적으로 동작을 숨깁니다. 코드를 읽는 다른 사람이 파악하는 것을 더 어렵게하죠.

그냥 로직을 public method 안에 두어도 충분합니다.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/domain/Issue.java:
```diff
@@ -3,89 +3,167 @@
+        assignees.addAll(newAssignees);
     }
 
+    /**
+     * 이슈의 상태를 변경
+     */
```
이런 주석은 거의 의미가 없습니다. 메서드 이름만 봐도 알 수 있습니다.
차라리 사용법을 적는게 낫죠.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/domain/Issue.java:
```diff
@@ -3,89 +3,167 @@
-        lastModifiedAt = LocalDateTime.now();
     }
 
+    /**
+     * 리스트에 없는 회원을 기존의 할당자에서 제거
+     */
+    private void removeAssigneesNotIn(List<Member> changeMembers) {
+        this.assignees.removeIf(
+                assignee -> !assignee.isMemberBelongTo(changeMembers)
+        );
+    }
+
+    /**
+     * 기존에 추가되지 않은 회원들로 IssueAssignee 생성
+     */
+    private List<IssueAssignee> createNewAssignees(List<Member> changeMembers) {
+        return changeMembers.stream()
+                .map(newMember -> new IssueAssignee(newMember, this))
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * 리스트에 존재하는 할당자가 이미 존재하면 changeMembers에서 제거
+     */
+    private void removeDuplicateMemberFrom(List<Member> changeMembers) {
+        List<Member> beforeMembers = assignees.stream()
+                .map(IssueAssignee::getMember)
+                .collect(Collectors.toList());
+        changeMembers.removeIf(beforeMembers::contains);
+    }
```
예전에 리뷰해드린게 기억이 나실진 모르겠는데, `private` 메서드는 사용되는 `public` 메서드 아래에 위치시켜주는게 가독성에 좋습니다.

ttasjwi replied:
저희도 코드를 다시 확인해보니 기능구현에 급급하여 메서드 순서 조정을 나중으로 미룬 경우가 많았고 불만족스러운 것들을 몇 개 확인했는데, 이들과 더불어 다음 코드에서 반영하겠습니다!

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/domain/Issue.java:
```diff
@@ -3,89 +3,167 @@
+        this.issueLabels.removeIf(
+                issueLabel -> !issueLabel.isLabelBelongTo(changeLabels)
+        );
+    }
```
비슷한 코드죠? 메서드로 추출할건 이 메서드가 아니라 다른게 아닐까요?

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/domain/Issue.java:
```diff
@@ -3,89 +3,167 @@
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * 제목 변경
+     */
+    public void changeTitle(String title) {
+        this.title = title;
+    }
+
+    /**
+     * 본문 변경
+     */
+    public void changeContent(String content) {
+        this.content = content;
+    }
+
+    /**
+     * 마일스톤 변경
+     */
+    public void changeMilestone(Milestone updateMilestone) {
+        this.milestone = milestone;
+    }
```
👍

ttasjwi replied:
필드의 변경에 set...을 두면 의미가 불분명하기 떄문에, 변경의 이유를 메서드 명에 표현하였습니다!

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/domain/IssueAssignee.java:
```diff
@@ -26,18 +28,13 @@ public class IssueAssignee {
     private Issue issue;
 
-    private IssueAssignee(Member member, Issue issue) {
+    IssueAssignee(Member member, Issue issue) {
```
왜 접근 제어자를 없애는 선택을 해주셨나요?

ttasjwi replied:
외부에서 Assignee를 생성하지 않게 하고 issue.changeIssueAssignees 메서드를 통해 이슈 할당자 생성, 부착을 처리하기 위해서 메서드를 감췄습니다. 이들의 생성은 Issue라는 도메인에 종속적이여서 Issue를 거쳐서만 생성될 수 있게 하고 싶었습니다.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/repository/issue/IssueRepositoryImpl.java:
```diff
@@ -20,7 +31,8 @@ public IssueRepositoryImpl(EntityManager em) {
 
-    public Optional<Issue> findDetailIssueById(Long issueId) {
+    @Override
+    public Optional<Issue> findByIdWithAuthorAndMilestone(Long issueId) {
```
이름 잘 지어주셨네용

ttasjwi replied:
무엇을 페치조인해오는 지 사용측에서 편할 것 같아서 이렇게 이름을 변경했습니다.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/repository/issue/IssueRepositoryImpl.java:
```diff
@@ -32,4 +44,82 @@ public Optional<Issue> findDetailIssueById(Long issueId) {
+                .execute();
+    }
+
+    @Override
+    public Optional<IssueDetailResponse> findIssueDetail(Long issueId) {
+
+        IssueDetailResponse issueDetailResponse = getIssueDetailResponse(issueId);
+
+        if (issueDetailResponse == null) {
+            return Optional.empty();
+        }
+
+        List<IssueDetailResponse.IssueAssigneeDto> assigneeDtos = findIssueAssigneeDtos(issueId);
+        List<LabelDto> labelDtos = findLabelDtos(issueId);
+        List<IssueDetailResponse.CommentListElement> commentDtos = findCommentDtos(issueId);
+
+        issueDetailResponse.getAssignees().addAll(assigneeDtos);
+        issueDetailResponse.getLabels().addAll(labelDtos);
+        issueDetailResponse.getComments().addAll(commentDtos);
+
+        return Optional.of(issueDetailResponse);
+    }
```
별로 좋은 형태는 아닌 것 같군요. 작성하시면서 어떤 느낌같은게 오진 않았나요?
`Repository` 의 목적은 무엇일까요?

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/service/issue/IssueCommandService.java:
```diff
@@ -69,13 +73,27 @@ private Issue findIssue(Long issueId) {
+    public void updateIssue(Long issueId, IssueUpdateRequest updateRequest) {
+        Issue issue = issueRepository.findByIdWithAuthorAndMilestone(issueId)
+                .orElseThrow(() -> new IssueNotFoundException("일치하는 식별자의 이슈를 찾을 수 없습니다."));
+        Milestone updateMilestone = findMilestone(updateRequest.getMilestoneId());
```
`issueCommandService`인데, `milestone` 을 조회하는 메서드가 있네요. 뭔가 의아하지 않나요?

ttasjwi replied:
조회를 다른 서비스를 통해 하는건 어떻게 생각하시나요?
저는 이 부분이 궁금했는데, 서비스가 다른 서비스를 의존하는 것은 어떨지 궁금했습니다.

ksundong replied:
땃쥐는 어떻게 생각하시나요?

ttasjwi replied:
어제 휘랑 고민하고 service간 의존관계를 부여해봤습니다.

이를 통해 얻을 수 있는 장점이 매우 많았습니다. 관련 엔티티의 조회를 관련 쿼리서비스에서 집중적으로 책임지고,  커맨드 서비스는 동일 쿼리서비스를 의존하게 하게 하였는데요. 각 사비스에서 레포지토리를 의존하게했을 때는 각 서비스에서 조회에 대한 비슷한 검증 하는 코드를 계속 작성하는 문제가 있었고 전혀 관계없는 엔티티를 질의하는 문제가 있었습니다. 

서비스간 의존관계를 두니, 책임의 집중, 분리가 잘 되는 장점이 생겼습니다. 질의 시의 검증 등을 한 곳에서 책임지고 관리하므로 호출측에서는 그에 대해서 알 필요가 없어졌습니다. 물론 잘못된 의존관계를 맺어주면 순환참조가 발생할 여지가 있으니 이를 주의해야합니다.

ttasjwi replied:
그런데 이 부분에서 트랜잭션 중첩에 관한 개념을 학습해야겠다는 생각이 들었습니다. 학습할 거리를 많이 얻어가네요! 정말 감사합니다. 디온!

ttasjwi replied:
그런데 이 부분에서 트랜잭션 중첩에 관한 개념을 학습해야겠다는 생각이 들었습니다. 학습할 거리를 많이 얻어가네요! 정말 감사합니다. 디온!

ksundong replied:
네 스스로 느끼시는게 중요합니다.
제가 이게 맞다 말해도, 맞지 않을 수 있습니다. 코딩은 그런겁니다.
타인의 의견에 의존하지 않고, 본인의 주관을 뚜렷이 가지고 가시고, 다른 사람과 의견이 부딪히면 긍정적으로 검토해보시면 됩니다.
스스로 생각해보는 연습을 많이 해보시면 좋겠습니다.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/web/dto/issue/IssueDetailResponse.java:
```diff
@@ -0,0 +1,130 @@
+
+public class IssueDetailResponse {
+
+    private Long issueId;
+    private Long authorId;
+    private String authorName;
+    private String title;
+    private String content;
+    private final List<IssueAssigneeDto> assignees = new ArrayList<>();
+    private Long milestoneId;
+    private String milestoneName;
+    private boolean isOpened;
+    private LocalDateTime createdAt;
+    private LocalDateTime lastModifiedAt;
+    private final List<LabelDto> labels = new ArrayList<>();
+    private final List<CommentListElement> comments = new ArrayList<>();
```
저는 응답객체에는 `final` 을 붙여주는걸 권장합니다.
가급적이면 `final` 이 가능한 구간에는 다 붙여주세요.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/web/dto/issue/IssueDetailResponse.java:
```diff
@@ -0,0 +1,130 @@
+        this.lastModifiedAt = lastModifiedAt;
+    }
+
+    @Getter
+    public static class IssueAssigneeDto {
+
+        private Long assigneeId;
+        private String assigneeName;
+
+        @QueryProjection
+        public IssueAssigneeDto(Long assigneeId, String assigneeName) {
+            this.assigneeId = assigneeId;
+            this.assigneeName = assigneeName;
+        }
+    }
+
+    @Getter
+    public static class CommentListElement {
+
+        private Long commentId;
+        private Long commentAuthorId;
+        private String commentAuthorName;
+        private String commentContent;
+        private LocalDateTime commentCreatedTime;
+        private LocalDateTime commentLastModifiedTime;
+
+        @QueryProjection
+        public CommentListElement(Long commentId, Long commentAuthorId, String commentAuthorName, String commentContent, LocalDateTime commentCreatedTime, LocalDateTime commentLastModifiedTime) {
+            this.commentId = commentId;
+            this.commentAuthorId = commentAuthorId;
+            this.commentAuthorName = commentAuthorName;
+            this.commentContent = commentContent;
+            this.commentCreatedTime = commentCreatedTime;
+            this.commentLastModifiedTime = commentLastModifiedTime;
+        }
+
+    }
```
보통 이런경우 nested class는 권장하지 않습니다. 경험상 재사용이 되어서 분리되는 경우가 잦아서 그냥 별도 클래스로 쪼개는게 나은 것 같아요.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/web/dto/issue/IssueDetailResponse.java:
```diff
@@ -0,0 +1,130 @@
+
+    public String getTitle() {
+        return title;
+    }
+
+    public String getContent() {
+        return content;
+    }
+
+    public List<IssueAssigneeDto> getAssignees() {
+        return assignees;
+    }
+
+    public Long getMilestoneId() {
+        return milestoneId;
+    }
+
+    public String getMilestoneName() {
+        return milestoneName;
+    }
+
+    public boolean getIsOpened() {
+        return isOpened;
+    }
+
+    public LocalDateTime getCreatedAt() {
+        return createdAt;
+    }
+
+    public LocalDateTime getLastModifiedAt() {
+        return lastModifiedAt;
+    }
+
+    public List<LabelDto> getLabels() {
+        return labels;
+    }
+
+    public List<CommentListElement> getComments() {
+        return comments;
+    }
```
얘는 왜 getter를 직접 만들어주셨나요...?

ttasjwi replied:
boolean isOpened는 `@getter`로 getter생성시 isOpened로 getter가 생성되는 바람에 이렇게 처리했습니다.

ksundong replied:
아하 그랬군요.
Lombok의 네이밍 원칙이 primitive boolean을 메서드명으로 알 수 있게하자여서 그런 것 같네요.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/web/dto/label/LabelCreateRequest.java:
```diff
@@ -0,0 +1,26 @@
+    @NotBlank
+    private String name;
+    private String description;
+    private String colorCode;
```
`ArgumentResolver` 를 이용하면 좋지 않을까요?

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/web/IssueController.java:
```diff
@@ -64,26 +67,26 @@ public IssueUpdateResponse update(@RequestBody IssueUpdateRequest updateRequest)
-        issueCommandService.changeState(issueId, stateChangeRequest.getIsOpened());
-        Issue issue = issueQueryService.findIssue(issueId);
+        boolean isOpened = stateChangeRequest.getIsOpened();
+        log.info("Issue StateChange Request = {}", isOpened);
 
-        return IssueStateChangeResponse.create(issue);
+        issueCommandService.changeState(issueId, isOpened);
+        return new IssueStateChangeResponse(issueId, isOpened);
```
왜 이렇게 해줘야 하나요..?

ttasjwi replied:
변경 후 이슈 질의를 다시 할 필요가 있을까? 싶어서 요청의 값을 그대로 넣도록 변경했었습니다.

생각해보면, 실제 변경 이후 엔티티의 상태를 반환하는게 맞을것 같아서 다시 롤백시킬 예정입니다.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/web/IssueController.java:
```diff
@@ -64,26 +67,26 @@ public IssueUpdateResponse update(@RequestBody IssueUpdateRequest updateRequest)
      * 이슈 삭제
      */
     @DeleteMapping("/issue-tracker/api/issues/{issueId}")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
```
`NO_CONTENT` 가 맞나요?

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/web/LabelController.java:
```diff
@@ -0,0 +1,34 @@
+    @PostMapping("/issue-tracker/api/labels")
+    @ResponseStatus(HttpStatus.CREATED)
+    public LabelCreateResponse createLabel(@RequestBody LabelCreateRequest labelCreateRequest) {
+        Label label = labelCreateRequest.toEntity();
```
왜 여기서 `Entity` 로 만들게끔 해주셨나요?

ttasjwi replied:
요청 파라미터를 기반으로 엔티티를 생성하고 서비스에서는 단순히 등록만 해주려고 했었습니다.
근데 생각해보면 이슈를 생성하는 로직을 굳이 컨트롤러단에서 몰라도 되고, 서비스에서 책임지면 되므로 서비스에게 이 책임을 이동시키도록 하겠습니다.

ksundong reviewed (changes requested):
안녕하세요. 땃쥐 & 휘, 리뷰어 dion입니다.
미트볼을 추가했다는건 좋은 의미는 아닌데 말이죠 ㅋㅋㅋㅋㅋㅋㅋ

여러분의 코드는 스파게티정도는 아닙니다 ㅎㅎ

## 조언

- OAuth/JWT 꼭 해보세요.
- 예외처리는 굳이..? 싶긴 합니다. 나중에 해보셔도 돼요.
- 인터셉터도 꼭 해보세요.
- 테스트는 **무조건** 해보세요.

## 리뷰 요약

- `private` 메서드의 과도한 사용이 눈에 띕니다.
- 제 리뷰를 잘못 해석하고, 잘못 구현한 부분도 있는 것 같아요. 이전 PR 코멘트 참고해주세요.

고생하셨습니다. 이번에도 마찬가지로 일요일 9시에 머지할 예정입니다.
리뷰드린 내용 반영가능하신 곳 까지 반영해주세요.

ttasjwi committed:
[wnsxor1993#84] fix: MultipleIssueStateChangeResponse의 컴파일 에러 수정

- 변경 요청에서 재질의를 하는 과정을 없앴었다.
- 이 과정에서 IssueStateChangeResponse의 create 메서드를 제거했더니, 컴파일 에러가 발생함
- 이와 연관된 로직의 결함을 고민했는데, 이걸 고치기 전에 당장 컴파일 에러를 고치는게 중요하다 판단되서 당장 컴파일 에러가 나는 부분 먼저 고치기로 함
- 이후 리팩토링 예정

ttasjwi committed:
[wnsxor1993#85] refactor: Issue 관련 서비스, 레포지토리 의존관계 재정립

- IssueQueryService는 IssueRepository를 알고 있다.
- IssueCommandService는 IssueRepository, IssueQueryService를 알고 있다.
- Issue 조회 관련 책임을 IssueQueryService에 집중한다.

ttasjwi committed:
[wnsxor1993#85] refactor: Member 서비스, 리포지토리 의존관계 재정립

- Member 관련 질의는 MemberQueryService에서 집중하게 함

ttasjwi committed:
[wnsxor1993#85] refactor: Milestone 서비스, 리포지토리 의존관계 재정립

- Milestone 관련 질의는 MilestoneQueryService에서 집중하게 함

ttasjwi committed:
[wnsxor1993#85] refactor: Label 서비스, 리포지토리 의존관계 재정립

- Label 관련 질의는 LabelQueryService에서 집중하게 함

ttasjwi committed:
[wnsxor1993#86] refactor: IssueDetailResponse의 내부 클래스를 분리

- `IssueDetailResponse`의 내부 클래스인 `IssueAssigneeDto`를 따로 분리
- `IssueDetailResponse`의 내부 클래스인 `CommentListElement`를 따로 분리

ttasjwi committed:
[wnsxor1993#87] refactor: 수정된 Issue를 컨트롤러에서 재질의 하도록 변경

ttasjwi committed:
[wnsxor1993#88] refactor: label 생성을 컨트롤러에서 서비스로 옮김

ttasjwi committed:
[wnsxor1993#89] refactor: Issue에 IssueAssignee 추가 로직 리팩토링

기존에 IssueAssignee를 추가하거나, 수정하는 로직은 테스트하기 힘든 문제도 있고 코드 내부 복잡도가 너무 높았다. 불필요한 메서드가 공개할 여지가 있긴 하나, 테스트가 더 쉽도록 공개된 작은 메서드 하나하나의 정합성을 고려하여 메서드를 새로 작성했다.

- IssueAssignee의 동등성 기준을 member, issue로 함
- IssueAssignee의 public 생성자 열고, Issue 밖에서 수정할 수 있도록 함. (IssueAssignee의 테스트 편의성 증가)
- Issue의 Assginee 저장 컬렉션을 Set(구현체 HashSet)으로 변경(equals, hashCode 오버라이드 덕분에 Set에 중복 IssueAssignee를 저장하려 들면 실제로 저장되지 않는다.)
- List에 포함되지 않은 IssueAssignee를 제거하는 메서드를 외부에 공개

ttasjwi committed:
[wnsxor1993#89] refactor: Issue에 IssueLabel 추가 로직 리팩토링

- IssueAssignee 추가/수정과 구조적으로 동일한 방식으로 리팩토링

hwicode committed:
[wnsxor1993#90] refactor: 이슈 상세 조회 레포지토리 로직을 서비스로 분리

- `Issue`레포지토리에 있던 `findIssueDetail`메서드를 `IssueQueryService`로 분리함
- `Issue`레포지토리에 있던 `CommentDto` 조회 로직을 `CommentRepositoryImpl`로 분리함
- `IssueQueryService`가 `CommentQueryService`를 통해 `CommentDto`를 얻어오도록 함

ttasjwi committed:
[wnsxor1993#91] test: java 기본 String, HashSet 기능 테스트

ttasjwi committed:
[wnsxor1993#91] test: IssueAssignee의 동등성 테스트

ttasjwi committed:
[wnsxor1993#91] test: Issue에 issueAssignee 삽입 테스트

hwicode committed:
[wnsxor1993#91] test: IssueLabel의 동등성 테스트

hwicode committed:
[wnsxor1993#91] test: Issue에 issueLabel 삽입 테스트

ttasjwi committed:
[wnsxor1993#91] test: java 기본 Set의 기능 - contains 테스트

- 동등한 객체가 포함되어 있으면 true
- 동등한 객체가 포함되어 있지 않으면 false

ttasjwi committed:
[wnsxor1993#91] test: IssueAssignee - hasDifferentIssue 테스트

ttasjwi committed:
[wnsxor1993#91] test: IssueLabel - hasDifferentIssue 테스트

ttasjwi committed:
[wnsxor1993#91] test: Issue - removeAssigneesNotIn, removeIssueLabelsNotIn 테스트 추가

- 도중에 기존 코드의 오류를 확인하였고, 이를 수정했다.

ttasjwi committed:
[wnsxor1993#92] fix: 이슈 삭제 응답 Http 상태코드 재조정

- delete 요청 시 성공적으로 수행하면 200 응답
- 도중에 없는 식별자를 삭제요청하면 IssueNotFoundException 발생 - > ExceptionHandler에서 적절하게 상태코드 반환하게 할 예정

ttasjwi committed:
[wnsxor1993#85] refactor: 이슈를 Author, Milestone과 조회할 때 queryService를 거치도록 함

ttasjwi commented on BE/src/main/java/com/codesquad/issuetracker/service/issue/IssueCommandService.java:
```diff
@@ -63,15 +65,10 @@ private Member findMember(Long memberId) {
 
     public void changeState(Long issueId, boolean isOpened) {
-        Issue issue = findIssue(issueId);
+        Issue issue = issueQueryService.findIssueById(issueId);
```
Issue의 질의에 관한 부분은 QueryRepository에서 전부 맡도록 하고 CommandService가 해당 데이터 질의가 필요할 때는 QueryService에 요청하도록 했습니다.

ttasjwi commented on BE/src/main/java/com/codesquad/issuetracker/service/issue/IssueQueryService.java:
```diff
@@ -19,6 +19,11 @@ public class IssueQueryService {
 
     private final IssueRepository issueRepository;
 
+    public Issue findIssueById(Long issueId) {
```
조회 실패가 일어났을 때, 예외가 발생되도록 하는 로직은 질의를 책임지는 서비스단에서 집중하여 처리하게 하고 나머지 계층은 꺼내진 엔티티를 사용하도록 했습니다.

ksundong replied:
넵넵

hwicode committed:
[wnsxor1993#93] refactor:  Response 객체에서 가능한 구간은 final 붙혀줌

ttasjwi commented on BE/src/main/java/com/codesquad/issuetracker/service/label/LabelCommandService.java:
```diff
@@ -15,7 +16,8 @@ public class LabelCommandService {
     private final LabelRepository labelRepository;
 
-    public Long enrollLabel(Label label) {
+    public Long createLabel(LabelCreateRequest labelCreateRequest) {
```
웹 요청 Dto를 서비스의  파라미터로 넣는 형태가 옳은지는 약간 의문이긴 한데요.

이 방식을 실무에서 사용하는지 궁금하고... 만약 그렇게 안 쓰더라도
실무에서는 보통 파라미터가 복잡할 경우, 컨트롤러에서 이를 서비스 전용 DTO로 변환하고 인자로 넣어 사용하시는지 궁금합니다.

ksundong replied:
실무에서 쓰냐 안쓰냐는 별로 중요하지 않아요. 너무 실무중심으로 생각하는게 가장 큰 문제로보여요.
실무의 범위도 너무 넓을 뿐더러, 왜 그렇게 해야하는가가 '실무에서 그렇게 하니까'가 될 가능성이 너무 높아보입니다.
실무에선 모든게 가능합니다. getter setter 점철된 코드로 객체지향 없이 하는 것도 실무고, DDD에 기반한 설계에 TDD를 적용하는것도 가능합니다. 이런 질문은 '무의미한' 질문이 될 가능성이 높습니다.

질문을 정정해보죠.

> 저는 기존에 서비스에 Entity를 넘겨주도록 디자인 했는데요. 디온의 리뷰를 적용하고보니 웹 요청에 대한 DTO가 서비스의 파라미터가 되는 형태가 옳은지 의문이 듭니다.

그럼 제 답변은 이렇게 될겁니다.

왜 엔티티가 `Controller` layer에 등장해야 했나요? `Controller` 의 역할은 무엇인가요? 그리고 엔티티의 역할은 무엇인가요?

서비스의 재사용성을 생각하면, 서비스용 DTO로 변환해서 범용적으로 쓰이는 것이 더 맞을 것으로 보입니다. 아니면 도메인 객체를 도입하거나요.
다만, 우리가 지금 만드는 것은 엔드포인트가 한 곳인 서비스여서, 웹 컨트롤러의 DTO가 그대로 전달되어도 괜찮다고 봅니다.

ttasjwi commented on BE/src/main/java/com/codesquad/issuetracker/web/LabelController.java:
```diff
@@ -24,11 +23,8 @@ public class LabelController {
-
-        log.info("create Label = {}", label);
-
+        Long labelId = labelCommandService.createLabel(labelCreateRequest);
```
commandService의 반환 타입은 사이드 이펙트를 방지하기 위해(애초에 osiv를 껐기 때문에 외부에서 변경이 불가능하지만) 절대 엔티티를 반환하지 않게 했습니다.
반환하더라도 식별자 값이나 간단한 primitive 타입(boolean, int)들을 반환하게 했습니다.

ttasjwi commented on BE/src/main/java/com/codesquad/issuetracker/domain/IssueAssignee.java:
```diff
@@ -28,13 +31,18 @@ public class IssueAssignee {
     private Issue issue;
 
-    IssueAssignee(Member member, Issue issue) {
+    public IssueAssignee(Member member, Issue issue) {
```
디온의 말씀대로 객체 생성을 issue 안에서 하기보다 public으로 열어두고 외부에서 생성 뒤 issue에서 주입하는 식으로 처리해보니 장점이 더 많아진 것 같았습니다. 일단 객체를 생성해서 테스트가 가능하다는 것이 가장 핵심입니다. 테스트가 편한 코드를 작성하는게 중요하다는 것을 느꼈습니다.

ttasjwi commented on BE/src/main/java/com/codesquad/issuetracker/excption/InvalidIssueAssigneeAddException.java:
```diff
@@ -0,0 +1,11 @@
+/**
+ * 이슈에 IssueAssignee를 할당할 때 다른 Issue에 할당된 IssueAssignee를 추가할 때 발생하는 예외
+ */
+public class InvalidIssueAssigneeAddException extends BusinessException {
```
다른 Issue에 할당된 IssueAssignee를 추가하려는 상황은 web을 통해 service로 들어오는 일련의 과정을 확인하면 일어날 수 없는 일이지만 예외를 정의하고 검증로직을 추가했습니다.

ttasjwi commented on BE/src/main/java/com/codesquad/issuetracker/domain/IssueAssignee.java:
```diff
@@ -1,18 +1,21 @@
 @Entity
 @Getter
 @NoArgsConstructor(access = AccessLevel.PROTECTED)
+@EqualsAndHashCode(of = {"member", "issue"})
```
issueAssignee의 동등성을 회원, 이슈를 기준으로 하도록 정의했습니다.

ksundong replied:
왜 인지가 더 중요합니다. '왜' 인가요?

ksundong replied:
이게 엔티티가 엔티티의 역할과 도메인의 역할을 둘 다 가져서 약간의 혼란이 올 수도 있습니다. ㅎㅎ

ttasjwi replied:
정의 시 Issue의 참여자인 Issueassignee , issueLabel의 Set을 통한 중복 처리에서 편의성을 느껴서 재정의했습니다.이를 통한 장점이 너무 커서 이를 사용했습니다.

ksundong replied:
ㅎㅎ 맞아요. 사실 편의성의 문제가 아니라, 도메인의 정의상 그게 맞지요.
다만, 이게 순수한 엔티티였다면 어땠을까요?

엔티티와 도메인의 정의를 잘 생각해보세요. 😊

ttasjwi commented on BE/src/main/java/com/codesquad/issuetracker/domain/Issue.java:
```diff
@@ -25,7 +28,7 @@ public class Issue extends BaseTimeEntity {
 
     @OneToMany(mappedBy = "issue", cascade = CascadeType.ALL, orphanRemoval = true)
-    private final List<IssueAssignee> assignees = new ArrayList<>();
+    private final Set<IssueAssignee> assignees = new HashSet<>();
```
이슈의 회원을 변경하다보면 중복회원을 넣지 말아야하는데...
이 기준을 어떻게 하다가 동등성을 재정의하고 set으로 관리하도록 변경했습니다.

ksundong replied:
네 잘해주셨어요.

ttasjwi commented on BE/src/main/java/com/codesquad/issuetracker/domain/Issue.java:
```diff
@@ -61,50 +64,35 @@ public static Issue create(String title, String content, Member author, Mileston
-        );
+    public void addIssueAssignee(IssueAssignee issueAssignee) {
+        validateIssueAssignee(issueAssignee);
+        assignees.add(issueAssignee);
```
set에 동등한 객체를 add를 실행했을 때 기존 것을 덮어쓰는지 아무 것도 하지 않는 지 테스트를 진행했는데, 동등한 중복객체를 삽입하면 삽입 자체를 안 하더라구요. 그래서 set 구조로 바꾸고 add를 하기로 했습니다.

ttasjwi replied:
이에 대한 테스트 코드는 이후 커밋에 추가했습니다.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/domain/Issue.java:
```diff
@@ -1,91 +1,142 @@
-    public void changeIssueState(boolean isOpened) {
-        this.isOpened = isOpened;
-        lastModifiedAt = LocalDateTime.now();
+    private void validateIssueAssignee(IssueAssignee issueAssignee) {
+        if (issueAssignee.hasDifferentIssue(this)) {
+            throw new InvalidIssueAssigneeAddException("다른 이슈에 할당된 IssueAssignee를 추가할 수 없습니다.");
+        }
+    }
```
이 코드는 잘 이해가 안가네요.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/domain/IssueAssignee.java:
```diff
@@ -26,18 +31,18 @@ public class IssueAssignee {
+                .collect(Collectors.toList());
     }
 
+    public boolean hasDifferentIssue(Issue otherIssue) {
+        return !(issue.equals(otherIssue));
+    }
```
이 코드를 보면 문제를 알 수 있는데요.
member는 전혀 고려하지 않았습니다.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/repository/issue/IssueRepositoryImpl.java:
```diff
@@ -32,4 +38,51 @@ public Optional<Issue> findDetailIssueById(Long issueId) {
     }
+
+    @Override
+    public void updateBulkStates(List<Long> issueIds, boolean isOpened) {
```
여기서 말하는 `state`가 뭔가요?

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/service/issue/IssueCommandService.java:
```diff
@@ -25,57 +21,70 @@
+
+        Issue issue = Issue.create(title, content, author, milestone);
+
+        List<IssueAssignee> assignees = IssueAssignee.createIssueAssignees(issue, assigneeMembers);
+        List<IssueLabel> issueLabels = IssueLabel.createIssueLabels(issue, labels);
+
+        issue.addIssueAssignees(assignees);
+        issue.addIssueLabels(issueLabels);
```
만들고 추가하는 과정이 한 덩어리 일 수는 없나요?

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/service/issue/IssueQueryService.java:
```diff
@@ -21,39 +23,35 @@ public class IssueQueryService {
+        List<IssueLabelDto> issueLabelDtos = issueRepository.findIssueLabelDtosByIssueId(issueId);
+        List<CommentListElement> commentDtos = commentQueryService.findCommentDtosByIssueId(issueId);
 
-        List<IssueLabel> issueLabels = detailIssue.getIssueLabels();
-        for (IssueLabel issueLabel : issueLabels) {
-            issueLabel.getLabel();
-        }
+        issueDetailResponse.getAssignees().addAll(assigneeDtos);
+        issueDetailResponse.getLabels().addAll(issueLabelDtos);
+        issueDetailResponse.getComments().addAll(commentDtos);
```
아직도 이렇게 해줘야 하는 이유는 잘 이해가 안되네요.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/web/IssueController.java:
```diff
@@ -74,23 +78,21 @@ public IssueDetailResponse issueDetail(@PathVariable Long issueId) {
      */
     @DeleteMapping("/issue-tracker/api/issues/{issueId}")
-    public IssueDeleteResponse delete(@PathVariable Long issueId) {
+    @ResponseStatus(HttpStatus.OK)
```
이걸 써줄 필요가 있나요?

ksundong commented on BE/src/test/java/com/codesquad/issuetracker/domain/IssueTest.java:
```diff
@@ -0,0 +1,206 @@
+@Slf4j
+class IssueTest {
+
+    Member memberA = new Member("memberA", "1234!", "사용자1");
+    Member memberB = new Member("memberB", "1234!", "사용자2");
+    Label labelA = new Label("labelA", "widwdniw", LabelColor.create("FFFFFF"));
+    Label labelB = new Label("labelB", "widwdniw", LabelColor.create("AAAAAA"));
```
이렇게 자주 생성되는 테스트용 객체는 factory를 만들어주세요.

ksundong commented on BE/src/test/java/com/codesquad/issuetracker/domain/IssueTest.java:
```diff
@@ -0,0 +1,206 @@
+        issue.addIssueAssignee(issueAssigneeB);
+        IssueAssignee findIssueAssignee = issueAssignees.stream().findFirst().get();
+
+        assertThat(issueAssignees.size()).isEqualTo(1);
+        assertThat(issueAssigneeA).isNotSameAs(issueAssigneeB);
+        assertThat(findIssueAssignee).isSameAs(issueAssigneeA);
```
여러개의 `assertion` 은 `assertAll` 로 감싸주세요.

ksundong reviewed (approved):
고생하셨습니다. 😊
